File: LeetCode/Unknown/프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.java
class Solution {
    public int solution(int[] a) {
        int N = a.length;
        if (N <= 3) return N;
        
        int[] lefts = new int[N];
        int[] rights = new int[N];
        
        lefts[0] = a[0];
        rights[N-1] = a[N-1];
        
        for (int i = 1; i < N; i++) {
            lefts[i] = Math.min(lefts[i - 1], a[i]);
        }
        for (int i = N - 2; i >= 0; i--) {
            rights[i] = Math.min(rights[i + 1], a[i]);
        }
        
        int answer = 2;
        for (int i = 1; i < N - 1; i++) {
            if (lefts[i] >= a[i] || rights[i] >= a[i]) answer++;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/68646. 풍선 터트리기/풍선 터트리기.cpp
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<int> a) {
    int N = a.size();
    if (N <= 3) return N; 

    vector<int> leftMin(N, a[0]), rightMin(N, a[N-1]);
    for (int i = 1; i < N; i++)    leftMin[i] = min(leftMin[i-1], a[i]);
    for (int i = N-2; i >= 0; i--) rightMin[i] = min(rightMin[i+1], a[i]);

    int answer = 2; // 양 끝 풍선 항상 가능
    for (int i = 1; i < N-1; i++) {
        if (a[i] <= leftMin[i-1] || a[i] <= rightMin[i+1]) answer++;
    }
    
    return answer;
}